### PR TITLE
fix set dirty segment distribution to leader view

### PR DESCRIPTION
**Pull request by weiliu1031**
Signed-off-by: Wei Liu <wei.liu@zilliz.com>

issue: #26145

### CODE REVIEW

File: internal/querycoordv2/observers/leader_observer.go [hunk 1]
```diff
@@ -184,6 +184,12 @@ func (o *LeaderObserver) findNeedLoadedSegments(leaderView *meta.LeaderView, dis
 			}
 			loadInfo := utils.PackSegmentLoadInfo(resp, nil)
 
+			log.Debug("leader observer append a segment to set:",
+				zap.Int64("collectionID", leaderView.CollectionID),
+				zap.String("Channel", leaderView.Channel),
```
Comment (yah01) [inline internal/querycoordv2/observers/leader_observer.go:189]: `"Channel"` -> `"channel"`, or maybe next PR

File: internal/querycoordv2/observers/leader_observer.go [hunk 2]
```diff
@@ -210,11 +216,12 @@ func (o *LeaderObserver) findNeedRemovedSegments(leaderView *meta.LeaderView, di
...
 			continue
 		}
-		log.Debug("leader observer append a segment to remove:", zap.Int64("collectionID", leaderView.CollectionID),
-			zap.String("Channel", leaderView.Channel), zap.Int64("leaderViewID", leaderView.ID),
-			zap.Int64("segmentID", sid), zap.Bool("distMap_exist", ok),
-			zap.Bool("existInCurrentTarget", existInCurrentTarget),
-			zap.Bool("existInNextTarget", existInNextTarget))
+		log.Debug("leader observer append a segment to remove:",
+			zap.Int64("collectionID", leaderView.CollectionID),
+			zap.String("Channel", leaderView.Channel),
```
Comment (yah01) [inline internal/querycoordv2/observers/leader_observer.go:221]: same as line 189

File: internal/querycoordv2/observers/leader_observer.go [hunk 3]
```diff
@@ -184,6 +184,12 @@ func (o *LeaderObserver) findNeedLoadedSegments(leaderView *meta.LeaderView, dis
 			}
 			loadInfo := utils.PackSegmentLoadInfo(resp, nil)
 
+			log.Debug("leader observer append a segment to set:",
```
Comment (yah01) [inline internal/querycoordv2/observers/leader_observer.go:187]: Also the log will be formatted, so the ":" is not needed. Correct "append" to "appends"

File: internal/querynodev2/segments/manager.go [hunk 4]
```diff
@@ -132,26 +134,37 @@ func (mgr *segmentManager) Put(segmentType SegmentType, segments ...Segment) {
 	}
 
 	for _, segment := range segments {
-		if _, ok := targetMap[segment.ID()]; ok {
+		oldSegment, ok := targetMap[segment.ID()]
+
+		if ok && oldSegment.Version() >= segment.Version() {
```
Comment (yah01) [inline internal/querynodev2/segments/manager.go:139]: This won't happen, the segment manager stores only local segments

**Review by yah01 — COMMENTED** [review]
/lgtm

File: internal/querynodev2/segments/manager.go [hunk 5]
```diff
@@ -132,31 +136,69 @@ func (mgr *segmentManager) Put(segmentType SegmentType, segments ...Segment) {
...
+	if ok && segment.Version() >= newVersion {
+		log.Warn("Invalid segment distribution changed, skip it",
+			zap.Int64("segmentID", segment.ID()),
+			zap.Int64("oldVersion", segment.Version()),
+			zap.Int64("newVersion", newVersion),
+		)
+		return
+	}
+
+	segment.UpdateVersion(newVersion)
```
Comment (yah01) [inline internal/querynodev2/segments/manager.go:198]: this will panic if segment not exists

File: internal/querynodev2/segments/manager.go [hunk 6]
```diff
@@ -132,31 +136,69 @@ func (mgr *segmentManager) Put(segmentType SegmentType, segments ...Segment) {
...
+		targetMap = mgr.growingSegments
+	case SegmentTypeSealed:
+		targetMap = mgr.sealedSegments
+	default:
+		panic("unexpected segment type")
+	}
+
+	segment, ok := targetMap[segmentID]
+	if ok && segment.Version() >= newVersion {
+		log.Warn("Invalid segment distribution changed, skip it",
```
Comment (yah01) [inline internal/querynodev2/segments/manager.go:190]: the log should be changed

File: internal/querynodev2/segments/segment.go [hunk 7]
```diff
@@ -159,6 +126,10 @@ func (s *baseSegment) Version() int64 {
 	return s.version
 }
 
+func (s *baseSegment) UpdateVersion(value int64) {
```
Comment (yah01) [inline internal/querynodev2/segments/segment.go:129]: prefer parameter name `version`

File: internal/querynodev2/segments/segment.go [hunk 8]
```diff
@@ -73,39 +73,6 @@ type IndexedFieldInfo struct {
 	IndexInfo   *querypb.FieldIndexInfo
 }
 
-type Segment interface {
```
Comment (yah01) [inline internal/querynodev2/segments/segment.go:76]: Is this deleted on purpose?

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/querynodev2/segments/segment.go (hunk 8 above)
Comment (weiliu1031) [reply internal/querynodev2/segments/segment.go:76]: move to segment_interface.go,  make it easy to generate mock file

**Comment by weiliu1031** [comment]
rerun ut

File: internal/querynodev2/segments/manager.go [hunk 9]
```diff
@@ -132,31 +136,75 @@ func (mgr *segmentManager) Put(segmentType SegmentType, segments ...Segment) {
...
+	default:
+		panic("unexpected segment type")
+	}
+
+	segment, ok := targetMap[segmentID]
+	if !ok {
+		log.Warn("segment not exist, skip segment version change",
+			zap.Int64("segmentID", segmentID),
+			zap.Int64("newVersion", newVersion),
+		)
```
Comment (yah01) [inline internal/querynodev2/segments/manager.go:193]: return here

**Review by yah01 — CHANGES_REQUESTED** [review]

**Review by yah01 — APPROVED** [review]
/lgtm

**Pull request merged**